README.md
# cacheable-response

![Last version](https://img.shields.io/github/tag/Kikobeats/cacheable-response.svg?style=flat-square)
[![Build Status](https://img.shields.io/travis/Kikobeats/cacheable-response/master.svg?style=flat-square)](https://travis-ci.org/Kikobeats/cacheable-response)
[![Coverage Status](https://img.shields.io/coveralls/Kikobeats/cacheable-response.svg?style=flat-square)](https://coveralls.io/github/Kikobeats/cacheable-response)
[![Dependency status](https://img.shields.io/david/Kikobeats/cacheable-response.svg?style=flat-square)](https://david-dm.org/Kikobeats/cacheable-response)
[![Dev Dependencies Status](https://img.shields.io/david/dev/Kikobeats/cacheable-response.svg?style=flat-square)](https://david-dm.org/Kikobeats/cacheable-response#info=devDependencies)
[![NPM Status](https://img.shields.io/npm/dm/cacheable-response.svg?style=flat-square)](https://www.npmjs.org/package/cacheable-response)

> An HTTP compliant route path middleware for serving cache response with invalidation support.

## Why

Server Side Rendering (_SSR_) is a luxurious but necessary thing if you want to have a first class user experience.

The main issue of doing server-side things is the extra cost associated with dynamic things: The server will take CPU cycles to compute the value to be served, probably discarded in the next page reload, losing precious resources in the process.

Instead of serving a real time™ – and costly – response, we can say it is OK serving a pre-calculated response but much much cheaper.

That will save CPU cycles, saving them for things that really matters.

## Install

```bash
$ npm install cacheable-response --save
```

## Get Started

**cacheable-response** is a HTTP middleware for a serving pre-calculated response.

It's like a LRU cache but with all the logic necessary for auto-invalidate response copies and refresh them.

Imagine you are currently running an HTTP microservice to compute something heavy in terms of CPU

```js
const server = ({ req, res }) => {
  const data = doSomething(req)
  res.send(data)
}
```

To leverage caching capabilities, just you need to adapt your HTTP based project a bit for following **cacheable-response** interface

```js
const cacheableResponse = require('cacheable-response')

const ssrCache = cacheableResponse({
  get: ({ req, res }) => ({
    data: doSomething(req),
    ttl: 7200000 // 2 hours
  }),
  send: ({ data, res, req }) => res.send(data)
})
```

At least, **cacheable-response** needs two things:

- **get**: The method to be called for creating a fresh cacheable response associated with the current route path.
- **send**: It determines how the response should be rendered.

**cacheable-response** is _framework agnostic_: It could be used with any library that accepts `(request, response)` as input.

```js
const micro = require('micro')

/* Explicitly pass `cacheable-response` as server */
micro((req, res) => ssrCache({ req, res }))
```

That's include any express-like framework as well.

```js
const express = require('express')
const app = express()

/* Passing `cacheable-response` instance as middleware */
app.use((req, res) => ssrCache({ req, res }))
```

See more [examples](#examples).

At all times the cache status is reflected as `x-cache` headers in the response.

The first resource access will be a `MISS`.

```bash
HTTP/2 200
cache-control: public, max-age=7200, s-maxage=7200, stale-while-revalidate=300
ETag: "d-pedE0BZFQNM7HX6mFsKPL6l+dUo"
x-cache-status: MISS
x-cache-expired-at: 1h 59m 60s
```

Successive resource access under the `ttl` period returns a `HIT`

```bash
HTTP/2 200
cache-control: public, max-age=7170
cache-control: public, max-age=7170, s-maxage=7170, stale-while-revalidate=298
ETag: "d-pedE0BZFQNM7HX6mFsKPL6l+dUo"
x-cache-status: HIT
x-cache-expired-at: 1h 59m 30s
```

After `ttl` period expired, the cache will be invalidated and refreshed in the next request.

In case you need you can force invalidate a cache response passing `force=true` as part of your query parameters.

```bash
curl https://myserver.dev/user # MISS (first access)
curl https://myserver.dev/user # HIT (served from cache)
curl https://myserver.dev/user # HIT (served from cache)
curl https://myserver.dev/user?force=true # MISS (forcing invalidation)
```

## API

### cacheableResponse([options])

#### options

##### cache

Type: `boolean`<br/>
Default: `new Keyv({ namespace: 'ssr' })`

The cache instance used for backed your pre-calculated server side response copies.

The library delegates in [keyv](https://github.com/lukechilds/keyv), a tiny key value store with [multi adapter support](https://github.com/lukechilds/keyv#official-storage-adapters).

If you don't specify it, a memory cache will be used.

##### ttl

Type: `number`<br/>
Default: `7200000`

Number of milliseconds a cache response is considered valid.

After this period of time, the cache response should be refreshed.

This value can be specified as well providing it as part of [`.get`](#get) output.

If you don't provide one, this be used as fallback for avoid keep things into cache forever.

##### serialize

Type: `function`<br/>
Default: `JSON.stringify`

Set the serializer method to be used before compress.

##### deserialize

Type: `function`<br/>
Default: `JSON.parse`

Set the deserialize method to be used after decompress.

##### compress

Type: `boolean`<br/>
Default: `false`

Enable compress/decompress data using brotli compression format.

If you enable it, you need to an additional `iltorb` package:

```bash
npm install iltorb
```

##### revalidate

Type: `function`|`number`<br/>
Default: `ttl => ttl / 24`

Number of milliseconds that indicates grace period after response cache expiration for refreshing it in the background. the latency of the refresh is hidden from the user.

You can provide a function, it will receive [`ttl`](#ttl) as first parameter or a fixed value.

The value will be associated with [`stale-while-revalidate`](https://www.mnot.net/blog/2014/06/01/chrome_and_stale-while-revalidate) directive.

##### getKey

Type: `function`<br/>
Default: `(req, res) => normalizeUrl(req.url)__hash(options)`

It determinates how the cache key should be computed, receiving `req, res, options` as input.

##### get

_Required_<br/>
Type: `function`<br/>

The method to be called for creating a fresh cacheable response associated with the current route path.

```js
async function get ({ req, res }) {
  const data = doSomething(req, res)
  const ttl = 7200000 // 2 hours
  const headers = { userAgent: 'cacheable-response' }
  return { data, ttl, headers }
}
```

The method will received `({ req, res })` and it should be returns:

- **data** `string`: The content to be saved on the cache.
- **ttl** `number`: The quantity of time in milliseconds the content is considered valid on the cache. Don't specify it means use default [`ttl`](#ttl).
- **createdAt** `date`: The timestamp associated with the content (`Date.now()` by default).

Any other property can be specified and will passed to `.send`.

##### send

_Required_<br/>
Type: `function`<br/>

The method used to determinate how the content should be rendered.

```js
async function send ({ req, res, data, headers }) {
  res.setHeader('user-agent', headers.userAgent)
  res.send(data)
}
```

It will receive `({ req, res, data, ...props })` being `props` any other data supplied to `.get`.

## Pro-tip: Distributed cache with CloudFlare™️

> This content is not sponsored; Just I consider CloudFlare is doing a good job offering a cache layer as part of their free tier.

Imagine what could be better than having one cache layer?

Exactly, two cache layers.

If your server domain is connected with CloudFlare you can take advantage of having a distributed CDN that also caches your responses.

![](https://i.imgur.com/2BCHVzh.png)

For doing that, you need to setup a `Page Rule` over your domain specifing you want to enable cache. [Read more how to do that](https://support.cloudflare.com/hc/en-us/articles/115000150272-How-do-I-use-Cache-Everything-with-Cloudflare-).

Next time you query about a resource, a new `cf-cache-status` appeared as part of your headers response.

```bash
HTTP/2 200
cache-control: public, max-age=7200, s-maxage=7200, stale-while-revalidate=300
ETag: "d-pedE0BZFQNM7HX6mFsKPL6l+dUo"
x-cache-status: MISS
x-cache-expired-at: 1h 59m 60s
cf-cache-status: MISS
```

CloudFlare will [respect your `cache-control` policy](https://support.cloudflare.com/hc/en-us/articles/202775670-How-Do-I-Tell-Cloudflare-What-to-Cache-), creating another caching layer reflected by `cf-cache-status`

```bash
HTTP/2 200
cache-control: public, max-age=7200, s-maxage=7200, stale-while-revalidate=300
ETag: "d-pedE0BZFQNM7HX6mFsKPL6l+dUo"
x-cache-status: MISS
x-cache-expired-at: 1h 59m 60s
cf-cache-status: HIT
```

Note how in this second request `x-cache-status` is still
`MISS`.

That's because CloudFlare way for caching the content includes caching the response headers.

The headers associated with the cache copy will the headers from the first request. You need to look at `cf-cache-status` instead.

You can have a better overview of the percentage of success by looking your CloudFlare domain analytics

![](https://i.imgur.com/1Eg64YS.png)

## Examples

> Make a PR for adding your project!

### Express

- [`unavatar.now.sh`](https://github.com/Kikobeats/unavatar/blob/master/src/index.js#L12)
- [`html-microservice`](https://github.com/microlinkhq/html/blob/master/src/index.js#L9)

### Next.js

- [`ssr-caching`](https://github.com/zeit/next.js/tree/canary/examples/ssr-caching)

## Bibliography

- [Server rendered pages are not optional, by Guillermo Rauch](https://rauchg.com/2014/7-principles-of-rich-web-applications#server-rendered-pages-are-not-optional).
- [Increasing the Performance of Dynamic Next.JS Websites, by scale AI](https://scale.ai/blog/performance-on-next-js-websites).
- [The Benefits of Microcaching, by NGINX](https://www.nginx.com/blog/benefits-of-microcaching-nginx/).
- [Cache-Control for Civilians, by Harry Robert](https://csswizardry.com/2019/03/cache-control-for-civilians/)

## License

**cacheable-response** © [Kiko Beats](https://kikobeats.com), released under the [MIT](https://github.com/Kikobeats/cacheable-response/blob/master/LICENSE.md) License.<br/>
Authored and maintained by Kiko Beats with help from [contributors](https://github.com/Kikobeats/cacheable-response/contributors).

> [kikobeats.com](https://kikobeats.com) · GitHub [Kiko Beats](https://github.com/Kikobeats) · Twitter [@Kikobeats](https://twitter.com/Kikobeats)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.9.4](https://github.com/IsobarItaly/cacheable-response/compare/v1.9.3...v1.9.4) (2019-12-03)

### [1.9.3](https://github.com/IsobarItaly/cacheable-response/compare/v1.7.3...v1.9.3) (2019-12-03)

### Features

* Added option parameter to getKey function


### [1.9.2](https://github.com/Kikobeats/cacheable-response/compare/v1.9.1...v1.9.2) (2019-11-25)

### [1.9.1](https://github.com/Kikobeats/cacheable-response/compare/v1.9.0...v1.9.1) (2019-11-19)

## [1.9.0](https://github.com/Kikobeats/cacheable-response/compare/v1.7.5...v1.9.0) (2019-11-04)


### Features

* return 304 when If-None-Match matches ETag ([ca08d9b](https://github.com/Kikobeats/cacheable-response/commit/ca08d9b877600fbf2e4c2bb73a041676cd7002c2)), closes [#31](https://github.com/Kikobeats/cacheable-response/issues/31)
* return 304 when If-None-Match matches ETag ([#32](https://github.com/Kikobeats/cacheable-response/issues/32)) ([ef672a1](https://github.com/Kikobeats/cacheable-response/commit/ef672a17a77c927dcb5c344d5a578bfbfa9167a5))

## [1.8.0](https://github.com/Kikobeats/cacheable-response/compare/v1.7.5...v1.8.0) (2019-10-30)


### Features

* return 304 when If-None-Match matches ETag ([ca08d9b](https://github.com/Kikobeats/cacheable-response/commit/ca08d9b877600fbf2e4c2bb73a041676cd7002c2)), closes [#31](https://github.com/Kikobeats/cacheable-response/issues/31)
* return 304 when If-None-Match matches ETag ([#32](https://github.com/Kikobeats/cacheable-response/issues/32)) ([ef672a1](https://github.com/Kikobeats/cacheable-response/commit/ef672a17a77c927dcb5c344d5a578bfbfa9167a5))

### [1.7.5](https://github.com/Kikobeats/cacheable-response/compare/v1.7.4...v1.7.5) (2019-10-30)

### [1.7.4](https://github.com/Kikobeats/cacheable-response/compare/v1.7.3...v1.7.4) (2019-10-15)

### [1.7.3](https://github.com/Kikobeats/cacheable-response/compare/v1.7.2...v1.7.3) (2019-09-21)

### [1.7.2](https://github.com/Kikobeats/cacheable-response/compare/v1.7.1...v1.7.2) (2019-09-11)

### [1.7.1](https://github.com/Kikobeats/cacheable-response/compare/v1.7.0...v1.7.1) (2019-09-11)

## [1.7.0](https://github.com/Kikobeats/cacheable-response/compare/v1.6.12...v1.7.0) (2019-09-11)


### Build System

* update dependencies ([5c502e6](https://github.com/Kikobeats/cacheable-response/commit/5c502e6))


### Features

* add response to getKey attributes ([da25929](https://github.com/Kikobeats/cacheable-response/commit/da25929))



### [1.6.12](https://github.com/Kikobeats/cacheable-response/compare/v1.6.11...v1.6.12) (2019-06-26)


### Build System

* handle unexpected situations ([5dda36e](https://github.com/Kikobeats/cacheable-response/commit/5dda36e))



### [1.6.11](https://github.com/Kikobeats/cacheable-response/compare/v1.6.10...v1.6.11) (2019-06-19)


### Build System

* update meta ([52576f3](https://github.com/Kikobeats/cacheable-response/commit/52576f3))



### [1.6.10](https://github.com/Kikobeats/cacheable-response/compare/v1.6.9...v1.6.10) (2019-06-18)


### Build System

* explicit return ([69507e9](https://github.com/Kikobeats/cacheable-response/commit/69507e9))



### [1.6.9](https://github.com/Kikobeats/cacheable-response/compare/v1.6.8...v1.6.9) (2019-06-18)


### Build System

* exclude embed query parameter ([2f58d0d](https://github.com/Kikobeats/cacheable-response/commit/2f58d0d))



### [1.6.8](https://github.com/Kikobeats/cacheable-response/compare/v1.6.7...v1.6.8) (2019-06-18)


### Build System

* explicit return ([6c1c6a6](https://github.com/Kikobeats/cacheable-response/commit/6c1c6a6))
* remove explicit return ([dd8d826](https://github.com/Kikobeats/cacheable-response/commit/dd8d826))



### [1.6.7](https://github.com/Kikobeats/cacheable-response/compare/v1.6.6...v1.6.7) (2019-06-12)


### Build System

* exclude proxy from query parameters ([3067172](https://github.com/Kikobeats/cacheable-response/commit/3067172))



### [1.6.6](https://github.com/Kikobeats/cacheable-response/compare/v1.6.5...v1.6.6) (2019-06-04)


### Build System

* better debug ([da881d4](https://github.com/Kikobeats/cacheable-response/commit/da881d4))



### [1.6.5](https://github.com/Kikobeats/cacheable-response/compare/v1.6.4...v1.6.5) (2019-05-28)


### Build System

* better debug log ([497e953](https://github.com/Kikobeats/cacheable-response/commit/497e953))



### [1.6.4](https://github.com/Kikobeats/cacheable-response/compare/v1.6.3...v1.6.4) (2019-05-20)


### Build System

* change git-authors-cli position ([d1e9879](https://github.com/Kikobeats/cacheable-response/commit/d1e9879))



### [1.6.3](https://github.com/Kikobeats/cacheable-response/compare/v1.6.2...v1.6.3) (2019-05-17)


### Build System

* add debug ([5b7e429](https://github.com/Kikobeats/cacheable-response/commit/5b7e429))



### [1.6.2](https://github.com/Kikobeats/cacheable-response/compare/v1.6.1...v1.6.2) (2019-05-16)



### [1.6.1](https://github.com/Kikobeats/cacheable-response/compare/v1.6.0...v1.6.1) (2019-05-06)


### Bug Fixes

* condition ([c9f8761](https://github.com/Kikobeats/cacheable-response/commit/c9f8761))



# [1.6.0](https://github.com/Kikobeats/cacheable-response/compare/v1.5.2...v1.6.0) (2019-05-05)


### Bug Fixes

* **package:** update pretty-ms to version 5.0.0 ([63c6dab](https://github.com/Kikobeats/cacheable-response/commit/63c6dab))


### Features

* expose .getKey method ([3603e4d](https://github.com/Kikobeats/cacheable-response/commit/3603e4d)), closes [#11](https://github.com/Kikobeats/cacheable-response/issues/11) [#22](https://github.com/Kikobeats/cacheable-response/issues/22)



<a name="1.5.2"></a>
## [1.5.2](https://github.com/Kikobeats/cacheable-response/compare/v1.5.1...v1.5.2) (2019-04-11)


### Bug Fixes

* remove origin form redis key ([b15375b](https://github.com/Kikobeats/cacheable-response/commit/b15375b))



<a name="1.5.1"></a>

index.js
'use strict'

const debug = require('debug-logfmt')('cacheable-response')
const createCompress = require('compress-brotli')
const normalizeUrl = require('normalize-url')
const { parse } = require('querystring')
const prettyMs = require('pretty-ms')
const hash = require('object-hash')
const assert = require('assert')
const getEtag = require('etag')
const { URL } = require('url')
const Keyv = require('keyv')

function isEmpty (value) {
  return (
    value === undefined ||
    value === null ||
    (typeof value === 'object' && Object.keys(value).length === 0) ||
    (typeof value === 'string' && value.trim().length === 0)
  )
}

const _getKey = (req, res, options = {}) => {
  const url = new URL(req.url, 'http://localhost').toString()
  const { origin } = new URL(url)
  const baseKey = normalizeUrl(url, {
    removeQueryParameters: [
      'embed',
      'filter',
      'force',
      'proxy',
      'ref',
      /^utm_\w+/i
    ]
  })

  const firstPartKey = baseKey.replace(origin, '').replace('/?', '')
  const secondPartKey = hash(options)

  return `${firstPartKey}__${secondPartKey}`
}

const toSeconds = ms => Math.floor(ms / 1000)

const createSetHeaders = ({ revalidate }) => {
  return ({ res, createdAt, isHit, ttl, hasForce, etag }) => {
    // Specifies the maximum amount of time a resource
    // will be considered fresh in seconds
    const diff = hasForce ? 0 : createdAt + ttl - Date.now()
    const maxAge = toSeconds(diff)

    res.setHeader(
      'Cache-Control',
      `public, must-revalidate, max-age=${maxAge}, s-maxage=${maxAge}, stale-while-revalidate=${
        hasForce ? 0 : toSeconds(revalidate(ttl))
      }`
    )

    res.setHeader('X-Cache-Status', isHit ? 'HIT' : 'MISS')
    res.setHeader('X-Cache-Expired-At', prettyMs(diff))
    res.setHeader('ETag', etag)
  }
}

module.exports = ({
  cache = new Keyv({ namespace: 'ssr' }),
  compress: enableCompression = false,
  getKey = _getKey,
  get,
  send,
  revalidate = ttl => ttl * 0.8,
  ttl: defaultTtl = 7200000,
  ...compressOpts
} = {}) => {
  assert(get, '.get required')
  assert(send, '.send required')

  const setHeaders = createSetHeaders({
    revalidate: typeof revalidate === 'function' ? revalidate : () => revalidate
  })

  const { serialize, compress, decompress } = createCompress({
    enable: enableCompression,
    ...compressOpts
  })

  return async ({ req, res, ...opts }) => {
    const hasForce = Boolean(
      req.query ? req.query.force : parse(req.url.split('?')[1]).force
    )
    const key = getKey(req, res, opts)
    const cachedResult = await decompress(await cache.get(key))
    const isHit = !hasForce && cachedResult !== undefined && cachedResult !== null
    const result = isHit ? cachedResult : await get({ req, res, ...opts })

    if (!result) return

    const {
      etag: cachedEtag,
      ttl = defaultTtl,
      createdAt = Date.now(),
      data,
      ...props
    } = result

    const etag = cachedEtag || getEtag(serialize(data))
    const ifNoneMatch = req.headers['if-none-match']
    const isModified = etag !== ifNoneMatch

    debug({
      key,
      isHit,
      cachedResult: !isEmpty(cachedResult),
      result: !isEmpty(result),
      etag,
      ifNoneMatch
    })

    setHeaders({
      etag,
      res,
      createdAt,
      isHit,
      ttl,
      hasForce
    })

    if (!isModified) {
      res.statusCode = 304
      res.end()
      return
    }

    if (!isHit) {
      const payload = { etag, createdAt, ttl, data, ...props }
      const value = await compress(payload)
      await cache.set(key, value, ttl)
    }

    return send({ data, res, req, ...props })
  }
}

module.exports.getKey = _getKey
